clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -213,33 +238,50 @@ public void ensureActiveGroup() {
             needsJoinPrepare = false;
         }
 
+        if (heartbeatThread == null) {
+            heartbeatThread = new HeartbeatThread();
+            heartbeatThread.start();
+        }
+
         while (needRejoin()) {
             ensureCoordinatorReady();
 
             // ensure that there are no pending requests to the coordinator. This is important
             // in particular to avoid resending a pending JoinGroup request.
-            if (client.pendingRequestCount(this.coordinator) > 0) {
+            if (state == MemberState.LEAVING || client.pendingRequestCount(this.coordinator) > 0) {
```

Comment:
It may not be necessary now since we are keeping track of the state more explicitly. Previously this was needed to prevent sending a JoinGroup before we completed a rebalance in progress. I'll think about it.